File: iOS/samples/TutorialAVOSCloud/TutorialAVOSCloud/WallPicturesViewController.m
//
//  WallPicturesViewController.m
//  TutorialAVOSCloud
//
//

#import "WallPicturesViewController.h"
#import "UploadImageViewController.h"

#import <AVOSCloud/AVOSCloud.h>

#import "Constants.h"

@interface WallPicturesViewController () {

}

@property (nonatomic, retain) NSArray *wallObjectsArray;
@property (nonatomic, retain) UIActivityIndicatorView *activityIndicator;


-(void)getWallImages;
-(void)loadWallViews;
-(void)showErrorView:errorString;

@end

@implementation WallPicturesViewController

@synthesize wallObjectsArray = _wallObjectsArray;
@synthesize wallScroll = _wallScroll;
@synthesize activityIndicator = _loadingSpinner;

#pragma mark - View lifecycle
- (void)loadView {
    [super loadView];
    UIScrollView *scrollView = [[UIScrollView alloc] initWithFrame:super.view.frame];
    self.view = scrollView;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.title = @"Wall";
    self.view.backgroundColor = RGB(50, 50, 50);
    self.wallScroll = (UIScrollView *)self.view;
    
    UIBarButtonItem *item = [[UIBarButtonItem alloc] initWithTitle:@"Upload" style:UIBarButtonItemStylePlain target:self action:@selector(goUpload:)];
    self.navigationItem.rightBarButtonItem = item;
    
    item = [[UIBarButtonItem alloc] initWithTitle:@"Logout" style:UIBarButtonItemStylePlain target:self action:@selector(logoutPressed:)];
    self.navigationItem.leftBarButtonItem = item;
}

-(void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];

    //Clean the scroll view
    for (id viewToRemove in [self.wallScroll subviews]){
        if ([viewToRemove isMemberOfClass:[UIView class]])
            [viewToRemove removeFromSuperview];
    }

    //Reload the wall
    [self getWallImages];
    
}

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
    // Return YES for supported orientations
    return (interfaceOrientation == UIInterfaceOrientationPortrait);
}

- (NSUInteger)supportedInterfaceOrientations {
    return UIInterfaceOrientationMaskPortrait;
}

#pragma mark Wall Load
//Load the images on the wall
//Load the images on the wall
-(void)loadWallViews
{
    //Clean the scroll view
    for (id viewToRemove in [self.wallScroll subviews]){
        
        if ([viewToRemove isMemberOfClass:[UIView class]])
            [viewToRemove removeFromSuperview];
    }
    
    
    //For every wall element, put a view in the scroll
    int originY = 10;
    
    for (AVObject *wallObject in self.wallObjectsArray){
        
        
        //Build the view with the image and the comments
        UIView *wallImageView = [[UIView alloc] initWithFrame:CGRectMake(10, originY, self.view.frame.size.width - 20 , 300)];
        
        //Add the image
        AVFile *image = (AVFile *)[wallObject objectForKey:KEY_IMAGE];
        UIImageView *userImage = [[UIImageView alloc] initWithImage:[UIImage imageWithData:image.getData]];
        userImage.frame = CGRectMake(0, 0, wallImageView.frame.size.width, 200);
        userImage.contentMode = UIViewContentModeScaleAspectFit;
        [wallImageView addSubview:userImage];
        
        //Add the info label (User and creation date)
        NSDate *creationDate = wallObject.createdAt;
        NSDateFormatter *df = [[NSDateFormatter alloc] init];
        [df setDateFormat:@"HH:mm dd/MM yyyy"];
        
        UILabel *infoLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 210, wallImageView.frame.size.width,15)];
        infoLabel.text = [NSString stringWithFormat:@"Uploaded by: %@, %@", [wallObject objectForKey:KEY_USER], [df stringFromDate:creationDate]];
        infoLabel.font = [UIFont fontWithName:@"Arial-ItalicMT" size:9];
        infoLabel.textColor = [UIColor whiteColor];
        infoLabel.backgroundColor = [UIColor clearColor];
        [wallImageView addSubview:infoLabel];
        
        //Add the comment
        UILabel *commentLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 240, wallImageView.frame.size.width, 15)];
        commentLabel.text = [wallObject objectForKey:KEY_COMMENT];
        commentLabel.font = [UIFont fontWithName:@"ArialMT" size:13];
        commentLabel.textColor = [UIColor whiteColor];
        commentLabel.backgroundColor = [UIColor clearColor];
        [wallImageView addSubview:commentLabel];
        
        [self.wallScroll addSubview:wallImageView];
        
        
        originY = originY + wallImageView.frame.size.width + 20;
        
    }
    
    //Set the bounds of the scroll
    self.wallScroll.contentSize = CGSizeMake(self.wallScroll.frame.size.width, originY);
    
    //Remove the activity indicator
    [self.activityIndicator stopAnimating];
    [self.activityIndicator removeFromSuperview];
}



#pragma mark Receive Wall Objects

//Get the list of images
-(void)getWallImages
{
    //Prepare the query to get all the images in descending order
    AVQuery *query = [AVQuery queryWithClassName:WALL_OBJECT];
    [query orderByDescending:KEY_CREATION_DATE];
    [query findObjectsInBackgroundWithBlock:^(NSArray *objects, NSError *error) {
        
        if (!error) {
            //Everything was correct, put the new objects and load the wall
            self.wallObjectsArray = nil;
            self.wallObjectsArray = [[NSArray alloc] initWithArray:objects];
            
            [self loadWallViews];
            
        } else {
            //Remove the activity indicator
            [self.activityIndicator stopAnimating];
            [self.activityIndicator removeFromSuperview];
            
            //Show the error
            NSString *errorString = [[error userInfo] objectForKey:@"error"];
            [self showErrorView:errorString];
        }
    }];

}


#pragma mark Actions


-(IBAction)logoutPressed:(id)sender
{
    [self.navigationController popViewControllerAnimated:YES];
}

-(void)goUpload:(id)sender
{
    UploadImageViewController *vc = [[UploadImageViewController alloc] init];
    [self.navigationController pushViewController:vc animated:YES];
}

#pragma mark Error Alert

-(void)showErrorView:(NSString *)errorMsg{
    
    UIAlertView *errorAlertView = [[UIAlertView alloc] initWithTitle:@"Error" message:errorMsg delegate:nil cancelButtonTitle:@"Ok" otherButtonTitles:nil, nil];
    [errorAlertView show];
}


@end

File: iOS/samples/TutorialAVOSCloud-Swift copy/TutorialAVOSCloud-Swift/WallPicturesViewController.swift
//
//  WallPicturesViewController.swift
//  TutorialAVOSCloud-Swift
//
//  Created by Qihe Bian on 6/10/14.
//  Copyright (c) 2014 AVOS. All rights reserved.
//

import UIKit

let imageViewTag:Int = 1001
let infoViewTag:Int = 1002
let commentViewTag:Int = 1003

class WallPicturesViewController: UITableViewController {
    var wallObjectsArray:NSArray?
    var activityIndicator:UIActivityIndicatorView?
    
    convenience init() {
        self.init(style: UITableViewStyle.Plain)
    }
    
    init(style: UITableViewStyle) {
        super.init(style: style)
        // Custom initialization
    }

    init(nibName nibNameOrNil: String!, bundle nibBundleOrNil: NSBundle!) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()

        self.title = "Wall"
        self.view.backgroundColor = RGB(50, 50, 50)
        self.tableView.separatorStyle = UITableViewCellSeparatorStyle.None
        self.tableView.registerClass(UITableViewCell.self, forCellReuseIdentifier: "reuseIdentifier")
        
        var item:UIBarButtonItem = UIBarButtonItem(title: "Upload", style: UIBarButtonItemStyle.Plain, target: self, action: "goUpload:")
        self.navigationItem.rightBarButtonItem = item
        
        item = UIBarButtonItem(title: "Logout", style: UIBarButtonItemStyle.Plain, target: self, action: "logoutPressed:")
        self.navigationItem.leftBarButtonItem = item
    }

    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        self.getWallImages()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    // #pragma mark - Table view data source

    override func numberOfSectionsInTableView(tableView: UITableView!) -> Int {
        // Return the number of sections.
        return 1
    }

    override func tableView(tableView: UITableView!, numberOfRowsInSection section: Int) -> Int {
        // Return the number of rows in the section.
        if(self.wallObjectsArray){
            return self.wallObjectsArray!.count
        } else {
            return 0
        }
    }

    override func tableView(tableView: UITableView!, heightForRowAtIndexPath indexPath: NSIndexPath!) -> CGFloat {
        return 300
    }
    
    override func tableView(tableView: UITableView!, cellForRowAtIndexPath indexPath: NSIndexPath!) -> UITableViewCell! {
        var cell:UITableViewCell = tableView.dequeueReusableCellWithIdentifier("reuseIdentifier", forIndexPath: indexPath) as UITableViewCell
        var imageView:UIImageView? = cell.contentView.viewWithTag(imageViewTag) as? UIImageView
        if(imageView == nil) {
            var originY:Float = 10
            cell.backgroundColor = RGB(50, 50, 50)
            //Add the image
            imageView = UIImageView(frame: CGRectMake(0, 0, tableView.frame.size.width, 200))
            imageView!.frame = CGRectMake(0, 0, tableView.frame.size.width, 200);
            imageView!.contentMode = UIViewContentMode.ScaleAspectFit;
            imageView!.tag = imageViewTag
            cell.contentView.addSubview(imageView)
            
            //Add the info label (User and creation date)
            var label:UILabel = UILabel(frame: CGRectMake(0, 210, tableView.frame.size.width,15))
            label.font = UIFont(name: "Arial-ItalicMT", size: 9)
            label.textColor = UIColor.whiteColor()
            label.backgroundColor = UIColor.clearColor()
            label.tag = infoViewTag
            cell.contentView.addSubview(label)
            
            //Add the comment
            label = UILabel(frame: CGRectMake(0, 240, tableView.frame.size.width, 15))
            label.font = UIFont(name: "ArialMT", size: 13)
            label.textColor = UIColor.whiteColor()
            label.backgroundColor = UIColor.clearColor()
            label.tag = commentViewTag
            cell.contentView.addSubview(label)
            
        }
        var wallObject:AVObject = self.wallObjectsArray?.objectAtIndex(indexPath.row) as AVObject
        var image:AVFile = wallObject.objectForKey(KEY_IMAGE) as AVFile
        imageView!.image = UIImage(data: image.getData())
        
        var label:UILabel? = cell.contentView.viewWithTag(infoViewTag) as? UILabel
        var creationDate:NSDate = wallObject.createdAt
        var df:NSDateFormatter = NSDateFormatter()
        df.dateFormat = "HH:mm dd/MM yyyy"
        var u:NSString = wallObject.objectForKey(KEY_USER) as NSString
        var d:String = df.stringFromDate(creationDate)
        label!.text = "Uploaded by: \(u), \(d)"
        
        label = cell.contentView.viewWithTag(commentViewTag) as? UILabel
        label!.text = wallObject.objectForKey(KEY_COMMENT) as? NSString

        return cell
    }

    /*
    // Override to support conditional editing of the table view.
    override func tableView(tableView: UITableView?, canEditRowAtIndexPath indexPath: NSIndexPath?) -> Bool {
        // Return NO if you do not want the specified item to be editable.
        return true
    }
    */

    /*
    // Override to support editing the table view.
    override func tableView(tableView: UITableView?, commitEditingStyle editingStyle: UITableViewCellEditingStyle, forRowAtIndexPath indexPath: NSIndexPath?) {
        if editingStyle == .Delete {
            // Delete the row from the data source
            tableView.deleteRowsAtIndexPaths([indexPath], withRowAnimation: .Fade)
        } else if editingStyle == .Insert {
            // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
        }    
    }
    */

    /*
    // Override to support rearranging the table view.
    override func tableView(tableView: UITableView?, moveRowAtIndexPath fromIndexPath: NSIndexPath?, toIndexPath: NSIndexPath?) {

    }
    */

    /*
    // Override to support conditional rearranging of the table view.
    override func tableView(tableView: UITableView?, canMoveRowAtIndexPath indexPath: NSIndexPath?) -> Bool {
        // Return NO if you do not want the item to be re-orderable.
        return true
    }
    */

    /*
    // #pragma mark - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepareForSegue(segue: UIStoryboardSegue?, sender: AnyObject?) {
        // Get the new view controller using [segue destinationViewController].
        // Pass the selected object to the new view controller.
    }
    */
    
    //Get the list of images
    func getWallImages() {
        //Prepare the query to get all the images in descending order
        var query:AVQuery = AVQuery(className: WALL_OBJECT)
        query.orderByDescending(KEY_CREATION_DATE)
        query.findObjectsInBackgroundWithBlock ({
            objects, error in
            if (!error) {
                //Everything was correct, put the new objects and load the wall
                self.wallObjectsArray = objects
                
                self.tableView.reloadData()
                
            } else {
                //Remove the activity indicator
                self.activityIndicator!.stopAnimating()
                self.activityIndicator!.removeFromSuperview()
                
                //Show the error
                var errorString:NSString = error.userInfo.objectForKey("error") as NSString
                self.showErrorView(errorString)
            }
            })
        
    }
    
    func logoutPressed(sender:AnyObject) {
        self.navigationController.popViewControllerAnimated(true)
    }
    
    func goUpload(sender:AnyObject) {
        var vc:UploadImageViewController = UploadImageViewController()
        self.navigationController.pushViewController(vc, animated: true)
    }
    
    
    func showErrorView(errorMsg:NSString) {
        
        var errorAlertView:UIAlertView = UIAlertView(title: "Error", message: errorMsg, delegate: nil, cancelButtonTitle: "OK")
        errorAlertView.show()
    }
}
